File: expansion_board/dht11/dht11.py
###############################################
#
#  file: dht11.py
#  update: 2024-08-10
#  usage: 
#      sudo python dht11.py
#
###############################################

import os
import struct
import time

# 打开设备节点
dht11_fd = os.open('/dev/dht11', os.O_RDONLY)

def main():
    try:
        while True:
            
            # 从设备节点读取6个字节的数据
            data = os.read(dht11_fd, 6)

            # 解析数据并打印
            print(f"Temperature: {data[2]}.{data[3]}℃, Humidity: {data[0]}.{data[1]}")

            time.sleep(1)

    except Exception as e:
        
        print("exit...：", e)
    
    finally:
        
        os.close(dht11_fd)
        
if __name__ == "__main__":  
    main()

File: expansion_board/dht11/dht11.c
/*
*
*   file: dht11.c
*   date: 2024-08-06
*   usage: 
*       sudo gcc -o dht11 dht11.c
*       sudo ./dht11
*
*/

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <sys/ioctl.h>
#include <poll.h>
#include <stdint.h>

#define DEV_NAME "/dev/dht11"

int main(int argc, char **argv)
{
	int fd;
	int ret;
    uint8_t data[6];
	
	/* open dev */
	fd = open(DEV_NAME, O_RDWR);
	if (fd < 0)
	{
		printf("can not open file %s, %d\n", DEV_NAME, fd);
		return -1;
	}

	while(1)
	{
		/* read date from dht11 */
		ret = read(fd, &data, sizeof(data));	
		if(ret)
		    printf("Temperature=%d.%d Humidity=%d.%d\n", data[2], data[3], data[0], data[1]);
        else
            printf("error reading!\n");
		
		sleep(1);
	}
	
	close(fd);
	
	return 0;
}


